pkg/genericapiserver/config.go

```diff
@@ -0,0 +1,467 @@
+	ServiceReadWritePort int
+
+	// The range of ports to be assigned to services with type=NodePort or greater
+	ServiceNodePortRange utilnet.PortRange
+
+	// Used to customize default proxy dial/tls options
+	ProxyDialer          apiserver.ProxyDialerFunc
+	ProxyTLSClientConfig *tls.Config
+
+	// Additional ports to be exposed on the GenericAPIServer service
+	// extraServicePorts is injectable in the event that more ports
+	// (other than the default 443/tcp) are exposed on the GenericAPIServer
+	// and those ports need to be load balanced by the GenericAPIServer
+	// service because this pkg is linked by out-of-tree projects
+	// like openshift which want to use the GenericAPIServer but also do
+	// more stuff.
+	ExtraServicePorts []api.ServicePort
+	// Additional ports to be exposed on the GenericAPIServer endpoints
+	// Port names should align with ports defined in ExtraServicePorts
+	ExtraEndpointPorts []api.EndpointPort
+
+	KubernetesServiceNodePort int
+
+	// EnableOpenAPISupport enables OpenAPI support. Allow downstream customers to disable OpenAPI spec.
+	EnableOpenAPISupport bool
+
+	// OpenAPIInfo will be directly available as Info section of Open API spec.
+	OpenAPIInfo spec.Info
+
+	// OpenAPIDefaultResponse will be used if an web service operation does not have any responses listed.
+	OpenAPIDefaultResponse spec.Response
+}
+
+func NewConfig(options *options.ServerRunOptions) *Config {
+	return &Config{
+		APIGroupPrefix:            options.APIGroupPrefix,
+		APIPrefix:                 options.APIPrefix,
+		CorsAllowedOriginList:     options.CorsAllowedOriginList,
+		AuditLogPath:              options.AuditLogPath,
+		AuditLogMaxAge:            options.AuditLogMaxAge,
+		AuditLogMaxBackups:        options.AuditLogMaxBackups,
+		AuditLogMaxSize:           options.AuditLogMaxSize,
+		EnableIndex:               true,
+		EnableLogsSupport:         options.EnableLogsSupport,
+		EnableProfiling:           options.EnableProfiling,
+		EnableSwaggerSupport:      true,
+		EnableSwaggerUI:           options.EnableSwaggerUI,
+		EnableUISupport:           true,
+		EnableWatchCache:          options.EnableWatchCache,
+		ExternalHost:              options.ExternalHost,
+		KubernetesServiceNodePort: options.KubernetesServiceNodePort,
+		MasterCount:               options.MasterCount,
+		MinRequestTimeout:         options.MinRequestTimeout,
+		PublicAddress:             options.AdvertiseAddress,
+		ReadWritePort:             options.SecurePort,
+		ServiceClusterIPRange:     &options.ServiceClusterIPRange,
+		ServiceNodePortRange:      options.ServiceNodePortRange,
+		EnableOpenAPISupport:      true,
+		OpenAPIDefaultResponse: spec.Response{
+			ResponseProps: spec.ResponseProps{
+				Description: "Default Response."}},
+		OpenAPIInfo: spec.Info{
+			InfoProps: spec.InfoProps{
+				Title:   "Generic API Server",
+				Version: "unversioned",
+			},
+		},
+	}
+}
+
+// SetDefaults fills in any fields not set that are required to have valid data.
```

Comment:
I'd like it documented why setting the defaults is a different operation
than construction.

On Tue, Aug 23, 2016 at 11:41 AM, David Eads <notifications@github.com>
wrote:

> In pkg/genericapiserver/config.go
> <https://github.com/kubernetes/kubernetes/pull/31247#discussion_r75926649>
> :
>
> > +		ServiceClusterIPRange:     &options.ServiceClusterIPRange,
> > +		ServiceNodePortRange:      options.ServiceNodePortRange,
> > +		EnableOpenAPISupport:      true,
> > +		OpenAPIDefaultResponse: spec.Response{
> > +			ResponseProps: spec.ResponseProps{
> > +				Description: "Default Response."}},
> > +		OpenAPIInfo: spec.Info{
> > +			InfoProps: spec.InfoProps{
> > +				Title:   "Generic API Server",
> > +				Version: "unversioned",
> > +			},
> > +		},
> > +	}
> > +}
> > +
> > +// SetDefaults fills in any fields not set that are required to have valid data.
>
> Explain why this isn't done in New()
>
> I don't think that New should mutate its inputs. Is your comment here
> saying that you think New should mutate its inputs or is your comment
> saying you'd like me to explain why I stopped calling this in New in the
> godoc?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31247/files/9f0608c1b455405ffbb87cd3b15f90e8f95cae16#r75926649>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglr8mWyRld8B2kY5oRUo2CUb3w0Igks5qiz7sgaJpZM4Jq_Kc>
> .
>